src/api/mod.rs

```diff
@@ -284,6 +284,9 @@ pub trait Central: Send + Sync + Clone {
 
     /// Returns a particular [`Peripheral`] by its address if it has been discovered.
     async fn peripheral(&self, address: BDAddr) -> Result<Self::Peripheral>;
+
+    /// Add a [`Peripheral`] from a MAC address without a scan result. Not supported on all Bluetooth systems.
+    async fn add_peripheral(&self, address: BDAddr) -> Result<Self::Peripheral>;
```

Comment:
What does it mean to add a peripheral? Add to what?

Reply:
So, let me explain the rationale behind this.

On Android, scanning for Bluetooth devices requires permission to access location information (though it looks like they're fixing this in Android 12.) To make this less cumbersome for developers, Android provides the [Companion Device Manager](https://developer.android.com/reference/android/companion/CompanionDeviceManager) API, which basically tells the Android OS "scan for Bluetooth devices yourself, prompt the user to select one, and tell me which one they selected." Unlike the scanning API, this does *not* require location permissions.

Once the app obtains this handle (which contains the MAC address information), either by prompting the user through the Companion Device Manager or by saving the MAC address and then retrieving it from storage later on, it can then connect to the device simply by providing the MAC address:

```java
BluetoothDevice device = BluetoothAdapter.getDefaultAdapter().getRemoteDevice("12:34:56:78:90:AB");
```

This new API is intended to make this functionality possible for btleplug clients on Android.

Reply:
This also becomes an interesting question in WebBluetooth with the upcoming getDevices() (or whatever it is) API, which I believe may allow similar?

Reply:
Would the app call the Companion Device Manager API directly, without btleplug involvement? And does this need to be a separate method from `peripheral`, just above? Either way, what you're really doing is getting a peripheral which the system already knows about, it's just a question of whether you are able to know about it beforehand. Perhaps they could be the same method?

Reply:
> Would the app call the Companion Device Manager API directly, without btleplug involvement?

Yes. The Companion Device Manager involves the use of an `Activity` and other things which are far above the level of btleplug.

My Android port also has an `Adapter::report_scan_result()` method, which takes an `android.bluetooth.le.ScanResult` (which can be obtained from the CDM) and applies it to the device as if the information in it had come from its own scan.

Reply:
> Either way, what you're really doing is getting a peripheral which the system already knows about, it's just a question of whether you are able to know about it beforehand. Perhaps they could be the same method?

In this branch, `Adapter::peripheral()` assumes the peripheral is already in the cache, while `Adapter::add_peripheral()` adds it if it's not already there. I suppose they could be merged into one method.

Reply:
> I suppose they could be merged into one method.

On second thought, I think it's better to be able to make the distinction between "get the peripheral only if it's already in the cache" and "add it by MAC address if it doesn't already exist". I'd prefer to keep `add_peripheral()` after all (I'm open to feedback on better names.)